### ISSUE TITLE
`Span` is really confusing

### ISSUE THREAD

**Issue by not-my-profile** [timothee-haudebourg/source-span#9]
The `Span` struct is arguably the most important type of this library. I however found it really confusing. I'll try to explain why.

The documentation starts with:

>     pub struct Span { /* fields omitted */ }
>A span points to a range of caracters between two cursor Position.

which sounds very reasonable. But then you find `Span::new` and everything falls apart:

`pub fn new(start: Position, last: Position, end: Position) -> Self`

Why are there suddenly three positions? After looking at both the documentation as well as the source code I still don't understand why you thought that this was a good idea. Any other source code spanning API does the reasonable thing and defines a span as two positions:

* [`rustc_span::SpanData`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_span/struct.SpanData.html) is defined by two positions
* [`proc_macro::Span`](https://doc.rust-lang.org/stable/proc_macro/struct.Span.html) has a `start` and an `end`
* [`codespan::Span`](https://docs.rs/codespan/0.11.1/codespan/struct.Span.html) has a `start` and an `end` position

To make things worse your naming is confusing as well:

```rust
/// The position of the first character in the span.
start: Position,

/// The last position in the span.
last: Position,

/// The position of the character directly following the span.
///
/// It is not included in the span.
end: Position,
```

So a span spans from `start` to `last`. That is completely unintuitive! I would expect either first - last or start - end ... but definitely not start - last. And it seems like it's not just me who's confused ... your documentation/code is also confused about what the third position `end` actually represents:

```rust
	/// Extends the span to include the rest of the line.
	///
	/// The end of the span will be placed at the begining of the next line.
	pub fn push_line(&mut self) {
		self.last = self.end;
		self.end = self.end.next_line();
	}
```

`Span::push_line` assumes that the `end` position points to the last character of the line. This assumption is not documented anywhere and directly conflicts with the claim that `end` is "the position of the character directly following the span.".

**tl;dr**: You really should remove the third position from `Span` and fix your API to be consistent.

**Comment by timothee-haudebourg** [comment]
Please try to not let too much frustration transpire in you comments. I am not as unreasonable as it seems, and things are as they are for a reason. Let me try to explain and we can then find some way to accommodate both of us.

This library was initially created to implement parsing libraries by providing a way to keep track of the position/span of each syntax token/node in a handy manner. In practice, users are *not expected to use the `Span::new` function* that exposes the problematic `end` parameter, but instead build `Span`s similarly to `String`s:

```rust
let mut token_content = String::default();
let mut token_span = Span::default();

while ... {
  match c {
    ... => {
      token_content.push(c);
      token_span.push(c, metrics);
    }
    ... => {
      token_content.clear();
      token_span.clear();
    }
  }
}
```

This is, in my opinion, an intuitive way to build spans in this situation which we can find in basically every tokenizer. However this also causes a slight implementation problem: `Span::clear` must transform the span so that it now starts and ends at the character that was directly following it. But because the span itself as no recollection of the actual characters it is composed of, we must keep track internally of the position of this next character. Hence the introduction of a third internal parameter, called `end`.

Now arguably:
- The third parameter is poorly named. I called it `end` for "historical" reasons that can only be confusing now. I guess `next` would be a better choice.
- What I explained here could be better explained in the documentation.

Do you still find it really confusing? If you really need to build a `Span` "by hand", which is not the recommended usage, then we can think of ways to accommodate both construction methods. One way would be to split this type into a `Span` type defined with two parameters, and some `SpanBuilder` type storing the third. But this would actually make the `Span` harder to use in some situations where keeping the `end` position was actually handy (I can think of some cases). So I won't do it unless it is really necessary.

**Comment by not-my-profile** [comment]
Sorry, I didn't mean to imply that you are unreasonable. Yes I gathered that this was meant to be convenient. It didn't work for me, because I was trying to integrate it into a parser that has tons of match blocks like:

```rs
match self.get_char() {
    '!' => { return State::Foo; }
    '/' => { return State::Bar; }
    c => {
        // I want to start the span here by doing e.g.
        self.current_name_span = self.current_pos.into();
        return State::Something;
    }
}
```

So the natural place to advance the current position is of course within `get_char()` the problem with that however is that the position already has been advanced when I know that I want to start a new span ... so I have to create a new span that starts one character before the current position ... which with your API currently is only possible with `Span::new` ... the very function users are not expected to use. So your convenience API actually gets in the way if you're unable to use it.

Anyway what bothers me most about the `Span` API are the completely undocumented assumptions it makes. `Span::new()` does not tell you that `end` is supposed to be at the end of the line. Actually in that case it doesn't even make sense that `Span` implements `Default` ... because for three `(0,0)` positions none of the `end` guarantees that `Span` tries to uphold are fulfilled.

**Comment by timothee-haudebourg** [comment]
You can solve your issue with the current API by reading each character in two steps: lookahead then consume:
```rust
let next_state = match self.peek_char() { // does not consume the char nor change current_name_span.
    '!' => State::Foo
    '/' => State::Bar
    c => {
        self.current_name_span.clear(); // start before c
        State::Something
    }
};

self.consume_char(); // actually consume the char and advance current_name_span.
next_state
```

That being said, it is true that having a default value for the third parameter makes no real sense. I'll try to see if I can break down the `Span` struct into more intuitive types in the next major version.

**Comment by timothee-haudebourg** [comment]
I'm currently preparing the next major release that should address your remarks. The `Span` definition has been simplified so that it contains only two positions (first and last). To resolve some inconsistencies, a `Span` can no longer be empty. On the other hand I defined a `Range` type, which is similar to a `Span` but can be empty (the last position is optional). The later can then be wrapped inside a `Lookahead<Range>` that also include the next character position. In this model, a `Span` can be constructed directly as you wanted, while `Lookahead<Range>` can be used to build spans character by character as I demonstrated above.

Would that satisfy you?

**Comment by not-my-profile** [comment]
Yes that does sound preferable to the current API. I want to note that I have since switched to the [codespan-reporting](https://github.com/brendanzab/codespan) library which allows spans [to be given as `core::ops::Range<usize>`](https://docs.rs/codespan-reporting/0.11.1/codespan_reporting/diagnostic/struct.Label.html#structfield.range) (that is just byte ranges within the source text).

I think you should really consider supporting `Range<usize>` because it lets parsers support your library without depending on it ... which imho is a major plus point. For example:

* the very popular `pulldown-cmark` library [provides spans as `Range<usize>`](https://docs.rs/pulldown-cmark/0.8.0/pulldown_cmark/struct.OffsetIter.html)
* my [html5tokenizer](https://docs.rs/html5tokenizer/0.4.0/html5tokenizer/spans/) library supports `Range<usize>`

Since I have mostly been filing issues here I do want to praise your documentation. It's very nice ... I particularly like your code examples ... codespan is really lacking those. When it however comes down to it API design is more important to me than documentation. (It's not just your current `Span` API that I dislike ... I also find it very weird that your [Formatter::render](https://docs.rs/source-span/latest/source_span/fmt/struct.Formatter.html#method.render) is taking an `Iterator<Item = Result<char, E>>` ... having to do `string.chars().map(Result::<char, ()>::Ok)` to pass a `String` or `&str` is just awkward).

**Comment by timothee-haudebourg** [comment]
I don't really like dealing with `Range<usize>` since it lacks the line/column information that is usually preferred to annotate AST nodes. However I see how it can be useful sometimes. I could add support for it through the `Layout` type that I already use to kind of go from `Range<usize>` to `Span`. I could enrich its API to explicitly accept the `Range<usize>` type and provide the `Range<usize>` <=> `Span` conversion in both directions.

**Comment by not-my-profile** [comment]
I strongly prefer `Range<usize>` because byte ranges are unambiguous. columns are ambiguous because there can be different tab widths, which you solve by requiring the API user to specify `DefaultMetrics` everywhere, e.g:
```rs
    let buffer = SourceBuffer::new(
        text.chars().map(Result::<char, ()>::Ok),
        Position::default(),
        DEFAULT_METRICS,
    );
    let formatted = fmt
        .render(
            buffer.iter(),
            Span::of_string(text, &DEFAULT_METRICS),
            &DEFAULT_METRICS,
        )
        .unwrap();
```

As an API user I find that cumbersome. As a parser library author I also don't want to deal with that.
